staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -62,10 +62,28 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
 	return a.namespaceScoped
 }
 
-func (customResourceDefinitionStorageStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
-}
-
-func (customResourceDefinitionStorageStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+func (a customResourceDefinitionStorageStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
+	// This applies defaults from the CRD schema.
+	// Validation errors, if any, are ignored. We do another validation pass here
+	// because we want defaulting to occur before the actual validation.
+	customResource := obj.(*unstructured.Unstructured)
+	customResourceValidator := a.validator.validator
+	result := customResourceValidator.Validate(customResource.UnstructuredContent())
+	if result.AsError() == nil {
+		result.ApplyDefaults()
```

Comment:
are errors possible here?

Reply:
no. The result object stores a list of closures setting the default values in the unstructured data type. The validation step before makes sure everything is valid. It constructs the defaulter on the fly.